File: src/listas/basica/vPython/List_mugre.py
from Node_mugre import Node

class List:
    def __init__(self):
        self.first = None

    def size(self):
        count = 0
        iterator = self.first
        while iterator:
            count += 1
            iterator = iterator.next
        return count

    def is_empty(self):
        return self.first is None

    def insert_end(self, value):
        new_node = Node(value)
        if not self.first:
            self.first = new_node
        else:
            iterator = self.first
            while iterator.next:
                iterator = iterator.next
            iterator.next = new_node

    def insert_front(self, value):
        new_node = Node(value)
        new_node.next = self.first
        self.first = new_node

    def delete(self):
        if self.first:
            if not self.first.next:
                self.first = None
            else:
                iterator = self.first
                while iterator.next.next:
                    iterator = iterator.next
                iterator.next = None

    def list_all(self):
        items = []
        iterator = self.first
        while iterator:
            items.append(iterator.value)
            iterator = iterator.next
        return items

File: src/listas/basica/vJavascript/List_mugre.js
const Node = require('./Node_mugre');

class List {
    constructor() {
        this.first = null;
    }

    size() {
        let count = 0;
        let current = this.first;
        while (current) {
            count++;
            current = current.next;
        }
        return count;
    }

    isEmpty() {
        return this.first === null;
    }

    insertEnd(value) {
        const newNode = new Node(value);
        if (!this.first) {
            this.first = newNode;
        } else {
            let current = this.first;
            while (current.next) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    insertFront(value) {
        const newNode = new Node(value);
        newNode.next = this.first;
        this.first = newNode;
    }

    delete() {
        if (!this.first) return;

        let current = this.first;
        let previous = null;
        while (current.next) {
            previous = current;
            current = current.next;
        }
        if (previous) {
            previous.next = null;
        } else {
            this.first = null;
        }
    }

    listAll() {
        let elements = [];
        let current = this.first;
        while (current) {
            elements.push(current.value);
            current = current.next;
        }
        return elements;
    }
}

module.exports = List;
